game/player.asm

```diff
@@ -80,64 +86,173 @@ Player_Input:
     pla
     pla
     pla
+    pla
     rts
 
 Player_UpBtn:
```

Comment:
I don't know if this will work, but here is how I would roughly organize it to keep your 8-bit operations together and not really have to _abuse_ the stack :D

```
Player_Input:
    jsr Player_UpBtn
    jsr Player_DnBtn
    jsr Player_LftBtn
    jsr Player_RhtBtn
    rts

Player_UpBtn:
    phx

    ; Load Input State Pointer
    lda player.input_obj_ptr, X
    tax

    ; Check if Up Button is pressed
    lda inputstate.upbtn, X
    and #1
    bne @Done

    ; Restore X pointing to the player object
    lda 1, S
    tax

    ; Now use X and Y index registers for oam object and char object pointers
    lda player.char_obj_ptr, X
    tay

    ; OAM object pointer is setup for any OAM calls now
    lda player.oam_obj_ptr, X
    tax

    ; Update the OAM object location by subtracting the speed from it
    ; This keeps our 8-bit operations all isolated
    A8
    sec
    lda oam_object.y, X
    sbc character_attr.speed, Y
    sta oam_object.y, X
    A16

    ; Update OAM (X pointer is already set)
    jsr OAM_MarkDirty

    @Done:
        plx
        rts
```